File: src/main/java/com/pyamsoft/pasterino/Pasterino.java
/*
 * Copyright 2016 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.pasterino;

import android.content.Context;
import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;
import com.pyamsoft.pasterino.dagger.DaggerPasterinoComponent;
import com.pyamsoft.pasterino.dagger.PasterinoComponent;
import com.pyamsoft.pasterino.dagger.PasterinoModule;
import com.pyamsoft.pydroid.lib.PYDroidApplication;

public class Pasterino extends PYDroidApplication implements IPasterino<PasterinoComponent> {

  private PasterinoComponent component;

  @NonNull @CheckResult public static IPasterino<PasterinoComponent> get(@NonNull Context context) {
    final Context appContext = context.getApplicationContext();
    if (appContext instanceof IPasterino) {
      return Pasterino.class.cast(appContext);
    } else {
      throw new ClassCastException("Cannot cast Application Context to IPasterino");
    }
  }

  @Override protected void onFirstCreate() {
    super.onFirstCreate();
    component = DaggerPasterinoComponent.builder()
        .pasterinoModule(new PasterinoModule(getApplicationContext()))
        .build();
  }

  @NonNull @Override public PasterinoComponent provideComponent() {
    if (component == null) {
      throw new NullPointerException("Pasterino component is NULL");
    }
    return component;
  }
}

File: app/src/main/java/com/pyamsoft/pasterino/Pasterino.kt
/*
 * Copyright 2020 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.pasterino

import android.app.Application
import androidx.annotation.CheckResult
import com.pyamsoft.pydroid.bootstrap.libraries.OssLibraries
import com.pyamsoft.pydroid.ui.PYDroid
import com.pyamsoft.pydroid.util.isDebugMode

class Pasterino : Application() {

    private val component by lazy {
        val url = "https://github.com/pyamsoft/pasterino"

        val provider = PYDroid.init(
            this,
            PYDroid.Parameters(
                viewSourceUrl = url,
                bugReportUrl = "$url/issues",
                version = BuildConfig.VERSION_CODE,
                termsConditionsUrl = TERMS_CONDITIONS_URL,
                privacyPolicyUrl = PRIVACY_POLICY_URL
            )
        )
        // Using pydroid-notify
        OssLibraries.usingNotify = true

        return@lazy DaggerPasterinoComponent.factory().create(
            isDebugMode(),
            this,
            provider.theming(),
            provider.imageLoader()
        )
    }

    override fun getSystemService(name: String): Any? {
        return PYDroid.getSystemService(name) ?: fallbackGetSystemService(name)
    }

    @CheckResult
    private fun fallbackGetSystemService(name: String): Any? {
        return if (PasterinoComponent::class.java.name == name) component else {
            super.getSystemService(name)
        }
    }

    companion object {

        const val PRIVACY_POLICY_URL =
            "https://pyamsoft.blogspot.com/p/pasterino-privacy-policy.html"
        const val TERMS_CONDITIONS_URL =
            "https://pyamsoft.blogspot.com/p/pasterino-terms-and-conditions.html"
    }
}
